# CryptoClustering - Challenge

In [71]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

### Load the data

In [72]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [73]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [74]:
# Validate data types and null values
df_market_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 41 entries, bitcoin to digibyte
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   price_change_percentage_24h   41 non-null     float64
 1   price_change_percentage_7d    41 non-null     float64
 2   price_change_percentage_14d   41 non-null     float64
 3   price_change_percentage_30d   41 non-null     float64
 4   price_change_percentage_60d   41 non-null     float64
 5   price_change_percentage_200d  41 non-null     float64
 6   price_change_percentage_1y    41 non-null     float64
dtypes: float64(7)
memory usage: 2.6+ KB


In [75]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [76]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_data_scaled = StandardScaler().fit_transform(df_market_data)

In [77]:
# Create a DataFrame with the scaled data
df_market_data_transformed = pd.DataFrame(df_market_data_scaled, columns=[
       'price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'])

# Copy the crypto names from the original data
df_market_data_transformed['coin_id'] = df_market_data.index

# Set the coinid column as index
df_market_data_transformed = df_market_data_transformed.set_index('coin_id')

# Display sample datad
df_market_data_transformed.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [78]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [79]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i)
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(df_market_data_transformed)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)

c:\Users\User\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


In [80]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k":k, "inertia":inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

# Review the DataFrame
df_elbow

k     inertia
0   1  287.000000
1   2  198.571818
2   3  123.190482
3   4   79.022435
4   5   64.008153
5   6   52.101806
6   7   43.786435
7   8   37.304991
8   9   33.027835
9  10   28.165433

In [81]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_plot = df_elbow.hvplot.line(
    x = "k",
    y = "inertia",
    title = "Elbow Curve",
    xticks = "k"
)

elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Best value for k is 4. k=4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [82]:
# Initialise the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [83]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_transformed)

KMeans(n_clusters=4)

In [84]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_4 = model.predict(df_market_data_transformed)

# Print the resulting array of cluster values.
k_4

array([3, 3, 0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 0, 3, 0, 3, 0, 0, 3, 0, 0, 3,
       0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 2, 3, 0, 0, 1, 0, 0, 0, 0])

In [85]:
# Create a copy of the DataFrame
df_market_data_predictions = df_market_data_transformed.copy()

In [86]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_predictions['CryptoClusters'] = k_4

# Display sample data
df_market_data_predictions.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  CryptoClusters  
coin_id                                                   
bitcoin                        -0.251637               3  
ethereum                       -0.199352               3  
tether                         -0.282061               0  
ripple                         -0.295546               0  
bitcoin-cash                   -0.270317               3

In [87]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
predictions_plot = df_market_data_predictions.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = "CryptoClusters",
    hover_cols=['coin_id'],
    title = "CryptoClusters"
)

predictions_plot

:NdOverlay   [CryptoClusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimise Clusters with Principal Component Analysis.

In [88]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [89]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_data_pca = pca.fit_transform(df_market_data_transformed)

# View the first five rows of the DataFrame. 
market_data_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [90]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [91]:
# Total explained variance
total_explained_variance = sum(pca.explained_variance_ratio_)
total_explained_variance

0.8950316570309842

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 
- Total explained variance = 89.50%
- PCA1 = 37.20%
- PCA2 = 34.70%
- PCA3 = 17.60%

In [92]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
df_market_data_pca = pd.DataFrame(
    market_data_pca, columns=["PCA1", "PCA2", "PCA3"]
)

# Copy the crypto names from the original data
df_market_data_pca['coin_id'] = df_market_data.index

# Set the coinid column as index
df_market_data_pca = df_market_data_pca.set_index('coin_id')

# Display sample data
df_market_data_pca.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [93]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1, 11))

In [94]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
for i in k_pca:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model_pca = KMeans(n_clusters=i)
# 2. Fit the model to the data using `df_market_data_pca`
    k_model_pca.fit(df_market_data_pca)
# 3. Append the model.inertia_ to the inertia list
    inertia_pca.append(k_model_pca.inertia_)

c:\Users\User\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


In [95]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k":k_pca, "inertia":inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)

# Review the DataFrame
df_elbow_pca

k     inertia
0   1  256.874086
1   2  168.811896
2   3   93.774626
3   4   49.665497
4   5   37.839466
5   6   27.618972
6   7   21.182776
7   8   17.283220
8   9   13.908202
9  10   10.484890

In [96]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_plot_pca = df_elbow_pca.hvplot.line(
    x = "k",
    y = "inertia",
    title="Elbow Curve PCA", 
    xticks=k
)

elbow_plot_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Best value for k is 4. k=4 (Same as previous model).

_  

* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it does not differ from the best k value found using the original data. Based on the elbow curve it is evident that at k=4 the line changes from more vertical to more horizontal.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [97]:
# Initialise the K-Means model using the best value for k
pca_model = KMeans(n_clusters=4)

In [98]:
# Fit the K-Means model using the PCA data
pca_model.fit(df_market_data_pca)

KMeans(n_clusters=4)

In [99]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4_pca = pca_model.predict(df_market_data_pca)

# Print the resulting array of cluster values.
k_4_pca

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 2, 0, 3, 3, 1, 3, 3, 3, 3])

In [100]:
# Create a copy of the DataFrame with the PCA data
df_market_data_pca_predictions = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_market_data_pca_predictions["CryptoClusters"] = k_4_pca

# Display sample data
df_market_data_pca_predictions

PCA1      PCA2      PCA3  CryptoClusters
coin_id                                                           
bitcoin              -0.600667  0.842760  0.461595               0
ethereum             -0.458261  0.458466  0.952877               0
tether               -0.433070 -0.168126 -0.641752               3
ripple               -0.471835 -0.222660 -0.479053               3
bitcoin-cash         -1.157800  2.041209  1.859715               0
binancecoin          -0.516534  1.388377  0.804071               0
chainlink            -0.450711  0.517699  2.846143               0
cardano              -0.345600  0.729439  1.478013               0
litecoin             -0.649468  0.432165  0.600303               0
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653               3
crypto-com-chain     -0.248198 -1.376252 -1.462026               3
usd-coin             -0.438408 -0.175337 -0.663388               3
eos                  -0.693425 -0.473815 -0.527597               3
monero                0.060499  2.909404  1.498571               0
tron                 -0.393352 -0.108192 -0.012756               3
tezos                -0.796176 -0.494409  1.082812               0
okb                   0.064075 -1.269825 -1.098829               3
stellar              -0.489015 -0.732719 -0.062543               3
cosmos               -0.306272  0.703415  1.714224               0
cdai                 -0.513528 -0.142802 -0.656566               3
neo                  -0.362120 -0.986914 -0.728752               3
wrapped-bitcoin      -0.604265  0.827398  0.439316               0
leo-token            -0.413296 -0.674115 -1.076628               3
huobi-token          -0.407483 -0.212507 -0.351426               3
nem                   0.608974  0.563532 -1.148742               3
binance-usd          -0.450211 -0.151019 -0.647401               3
iota                 -0.764665 -0.517886  0.204990               3
vechain              -0.556315 -1.938209 -1.261776               3
zcash                -0.425147  0.492976  1.058048               0
theta-token           2.676868 -0.013954 -1.965207               3
dash                 -0.613923 -0.479337  0.339565               3
ethereum-classic     -0.579924 -0.356334 -0.114942               3
ethlend               8.089018 -3.896891  2.301382               2
maker                -0.389045  0.165041  0.379414               0
havven                0.865762 -2.261882  0.275583               3
omisego               0.111675  0.428316 -1.205398               3
celsius-degree-token  4.792395  6.767679 -1.986985               1
ontology             -0.632355 -2.108117 -0.652227               3
ftx-token            -0.593142  0.021485  0.209911               3
true-usd             -0.458131 -0.135734 -0.635284               3
digibyte             -0.297910 -0.191126 -0.909602               3

In [101]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
predictions_plot_pca = df_market_data_pca_predictions.hvplot.scatter(
    x = "PCA1",
    y = "PCA2",
    by = "CryptoClusters",
    hover_cols=['coin_id'],
    title = "CryptoClusters PCA"
)

predictions_plot_pca

:NdOverlay   [CryptoClusters]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualise and Compare the Results

In this section, you will visually analyse the cluster analysis results by contrasting the outcome with and without using the optimisation techniques.

In [104]:
# Composite plot to contrast the Elbow curves
elbow_plot + elbow_plot_pca

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [105]:
# Composite plot to contrast the clusters
predictions_plot + predictions_plot_pca

:Layout
   .NdOverlay.I  :NdOverlay   [CryptoClusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [CryptoClusters]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analysing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The inertia of the clusters using PCA features is significantly lower than that of the original features. In the CryptoClusters PCA plot, the values within each cluster are more tightly grouped and easier to differentiate compared to the original data. The PCA plot also effectively distinguished the two outliers, celsius-degree-token and ethlend. This suggests that the accuracy remained good even with a 10% data loss.